internal/worker/worker.go

```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
+		go func(chunk []*big.Int) {
+			defer wg.Done()
+			resultsCh <- w.fetchLogsBatch(chunk)
+		}(chunk)
+	}
+	go func() {
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	results := make([]LogsFetchResult, 0, len(blockNumbers))
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
+	}
 
-func (w *Worker) fetchTraces(blockNumber *big.Int) ([]map[string]interface{}, error) {
-	var result []map[string]interface{}
-	err := w.rpc.RPCClient.Call(&result, "trace_block", fmt.Sprintf("0x%x", blockNumber))
-	return result, err
+	return results
 }
 
-func (w *Worker) queryRows() {
-	// TODO: Implement this
-	/*rows, err := w.storage.DBStorage.Query(context.Background(), "SELECT version()")
+func (w *Worker) fetchLogsBatch(blockNumbers []*big.Int) []LogsFetchResult {
+	batch := make([]rpc.BatchElem, len(blockNumbers))
+	logsResults := make([]LogsFetchResult, len(blockNumbers))
+
+	for i, blockNum := range blockNumbers {
+		logsResults[i] = LogsFetchResult{
+			BlockNumber: blockNum,
+		}
+		hexBlockNum := hexutil.EncodeBig(blockNum)
+		batch[i] = rpc.BatchElem{
+			Method: "eth_getLogs",
+			Args:   []interface{}{map[string]string{"fromBlock": hexBlockNum, "toBlock": hexBlockNum}},
+			Result: new([]map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		log.Error().Err(err).Msg("Error querying ClickHouse")
+		for i := range logsResults {
+			logsResults[i].Error = err
+		}
+		return logsResults
 	}
-	defer rows.Close()
 
-	for rows.Next() {
-		var version string
-		err = rows.Scan(&version)
-		if err != nil {
-			log.Error().Err(err).Msg("Error scanning version")
+	for i, elem := range batch {
+		if elem.Error != nil {
+			logsResults[i].Error = elem.Error
+		} else {
+			result := *elem.Result.(*[]map[string]interface{})
+			serializedLogs := SerializeLogs(w.rpc.ChainID, result)
+			logsResults[i].Logs = serializedLogs
 		}
-		log.Debug().Msgf("ClickHouse version: %s", version)
 	}
+	return logsResults
+}
+
+func (w *Worker) fetchTraces(blockNumbers []*big.Int) []TracesFetchResult {
+	if w.rpc.BlocksPerRequest.Traces >= len(blockNumbers) {
+		return w.fetchTracesBatch(blockNumbers)
+	}
+	chunks := divideIntoChunks(blockNumbers, w.rpc.BlocksPerRequest.Traces)
 
-	if err := rows.Err(); err != nil {
-		log.Error().Err(err).Msg("Error iterating rows")
+	log.Debug().Msgf("Fetching traces with %d chunks", len(chunks))
+
+	var wg sync.WaitGroup
+	resultsCh := make(chan []TracesFetchResult, len(chunks))
+
+	for _, chunk := range chunks {
+		wg.Add(1)
+		go func(chunk []*big.Int) {
+			defer wg.Done()
+			resultsCh <- w.fetchTracesBatch(chunk)
+		}(chunk)
+	}
+	go func() {
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	results := make([]TracesFetchResult, 0, len(blockNumbers))
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
+	}
+
+	return results
+}
+
+func (w *Worker) fetchTracesBatch(blockNumbers []*big.Int) []TracesFetchResult {
```

Comment:
this code looks duplicated across blocs, transactions and traces. is it possible to generalize it?

Reply:
Done